api/src/main/java/org/pmiops/workbench/db/dao/DataSetServiceImpl.java

```diff
@@ -216,101 +222,144 @@ public DataSet saveDataSet(
-          Optional.of(
-              dataSet.getValues().stream()
-                  .filter(valueSet -> valueSet.getDomain() == d)
-                  .collect(Collectors.toList()));
-      if (!valueSetOpt.isPresent()) {
-        continue;
-      }
+    return processDomains(domainList, dataSet, bigQuerySchemaConfig, includesAllParticipants,
+        conceptSetsSelected, cohortParametersBuilder, cohortQueries);
+  }
+
+  // TODO(jaycarlton) Convert to its own class or owherwise consolidate argument list
+  private Map<String, QueryJobConfiguration> processDomains(
+      List<Domain> domainList,
+      DataSetRequest dataSet,
+      CdrBigQuerySchemaConfig bigQuerySchemaConfig,
+      boolean includesAllParticipants,
+      List<ConceptSet> conceptSetsSelected,
+      Builder<String, QueryParameterValue> cohortParametersBuilder,
+      String cohortQueries) {
+
+    final ImmutableMap.Builder<String, QueryJobConfiguration> resultBuilder = new ImmutableBiMap.Builder<>();
+
+    for (Domain domain : domainList) {
+      final StringBuilder queryBuilder = new StringBuilder("SELECT ");
+
+      final List<DomainValuePair> valuePairsForCurrentDomain = dataSet.getValues().stream()
+          .filter(valueSet -> valueSet.getDomain() == domain)
+          .collect(Collectors.toList());
 
-      ValuesLinkingPair valuesLinkingPair = this.getValueSelectsAndJoins(valueSetOpt.get(), d);
-
-      query =
-          query
-              .concat(valuesLinkingPair.getSelects().stream().collect(Collectors.joining(", ")))
-              .concat(" ")
-              .concat(
-                  valuesLinkingPair.getJoins().stream()
-                      .distinct()
-                      .collect(Collectors.joining(" ")));
-
-      // CONCEPT SETS HERE:
-      conceptSetsSelected.stream()
-          .map(conceptSet -> conceptSet.getDomain())
-          .distinct()
-          .filter(cs -> d != Domain.PERSON)
-          .forEach(
-              domain -> {
-                if (conceptSetsSelected.stream()
-                        .filter(conceptSet -> conceptSet.getDomain() == domain)
-                        .mapToLong(conceptSet -> conceptSet.getConceptIds().size())
-                        .sum()
-                    == 0) {
-                  throw new BadRequestException("Concept Sets must contain at least one concept");
-                }
-              });
+      final ValuesLinkingPair valuesLinkingPair = getValueSelectsAndJoins(
+          valuePairsForCurrentDomain);
+
+      queryBuilder.append(String.join(", ", valuesLinkingPair.getSelects()))
+          .append(" ")
+          .append(formatValuesLinkingPair(valuesLinkingPair));
+
+      validateSelectedConceptSetsForDomain(conceptSetsSelected, domain);
 
       String conceptSetQueries =
           conceptSetsSelected.stream()
-              .filter(cs -> d == cs.getDomainEnum())
-              .filter(cs -> d != Domain.PERSON)
+              .filter(cs -> domain == cs.getDomainEnum())
```

Comment:
Could this be done as one line with an `&&`?

Reply:
The `(cs -> d != Domain.PERSON)` checks are getting moved external to this stream, since it doesn't depend on anything in the stream. I will probably need help with that logic though.